### PR TITLE
Added support for Synapse Private Link Hubs

### PR DESCRIPTION
**Pull request by renepajta**
# Change

Added support for Synapse Private Link Hubs. 

The module is able to deploy Synapse Private Link Hubs including Locks, RBACs and Private Endpoints.

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/workflows/ms.synapse.privatelinkhubs.yml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+          - major
+          - minor
+          - patch
+      customVersion:
+        description: 'Custom version to apply. Used only if higher than latest'
+        required: false
+        default: '0.0.1'
+  push:
+    branches:
+      - users/repajta/852_synapse_template
```
Comment (MrMCake) [inline .github/workflows/ms.synapse.privatelinkhubs.yml:26]:
```suggestion
      - main
```

**Review by MrMCake — CHANGES_REQUESTED** [review]
Please also add a pipeline in the .azuredevops folder

File: arm/Microsoft.Synapse/privateLinkHubs/.bicep/nested_rbac.bicep [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  'Managed Application Operator Role': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'c7393b34-138c-406f-901b-d8cf2b17e6ae')
+  'Managed Applications Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b9331d33-8a36-4f8c-b097-4f54124fdb44')
+  'Monitoring Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '749f88d5-cbae-40b8-bcfc-e573ddc772fa')
+  'Monitoring Metrics Publisher': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '3913510d-42f4-4e42-8a64-420c390055eb')
+  'Monitoring Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '43d0d8ad-25c7-4714-9337-8ba259a9fe05')
+  'Resource Policy Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '36243c78-bf99-498c-9df9-86d9f8d28608')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+resource synapsePrivateLinkHub 'Microsoft.Synapse/privateLinkHubs@2021-06-01' existing = {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/.bicep/nested_rbac.bicep:21]:
```suggestion
resource privateLinkHub 'Microsoft.Synapse/privateLinkHubs@2021-06-01' existing = {
```

File: arm/Microsoft.Synapse/privateLinkHubs/.bicep/nested_rbac.bicep [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  name: last(split(resourceId, '/'))
+}
+
+resource roleAssignment 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = [for principalId in principalIds: {
+  name: guid(synapsePrivateLinkHub.name, principalId, roleDefinitionIdOrName)
+  properties: {
+    roleDefinitionId: contains(builtInRoleNames, roleDefinitionIdOrName) ? builtInRoleNames[roleDefinitionIdOrName] : roleDefinitionIdOrName
+    principalId: principalId
+  }
+  scope: synapsePrivateLinkHub
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/.bicep/nested_rbac.bicep:31]:
```suggestion
  scope: privateLinkHub
```

File: arm/Microsoft.Synapse/privateLinkHubs/.parameters/parameters.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "value": "synplh0001"
+    },
+    "location": {
+      "value": "westeurope"
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/.parameters/parameters.json:9]: Please also add tests for at least RBAC to the parameter file. You can copy the parameters from for example StorageAccount

File: arm/Microsoft.Synapse/privateLinkHubs/.parameters/parameters.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "value": "synplh0001"
+    },
+    "location": {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/.parameters/parameters.json:8]: Remove the location here. It is fetched automatically

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 6]
```diff
@@ -0,0 +1,68 @@
+@description('Required. The name of the Private Link Hub.')
+param name string
+
+@description('Required. The geo-location where the resource lives.')
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:4]:
```suggestion
@description('Optional. The geo-location where the resource lives.')
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+@description('Optional. Specify the type of lock.')
+param lock string = 'NotSpecified'
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'')
+param roleAssignments array = []
+
+@description('Optional. Configuration Details for private endpoints.')
+param privateEndpoints array = []
+
+resource synapsePrivateLinkHub 'Microsoft.Synapse/privateLinkHubs@2021-06-01' = {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:24]:
```suggestion
resource privateLinkHub 'Microsoft.Synapse/privateLinkHubs@2021-06-01' = {
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// Resource Lock
+resource synapsePrivateLinkHub_lock 'Microsoft.Authorization/locks@2016-09-01' = if (lock != 'NotSpecified') {
+  name: '${synapsePrivateLinkHub.name}-${lock}-lock'
+  properties: {
+    level: lock
+    notes: (lock == 'CanNotDelete') ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: synapsePrivateLinkHub
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:37]:
```suggestion
  scope: privateLinkHub
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+param privateEndpoints array = []
+
+resource synapsePrivateLinkHub 'Microsoft.Synapse/privateLinkHubs@2021-06-01' = {
+  name: name
+  location: location
+  tags: tags
+}
+
+// Resource Lock
+resource synapsePrivateLinkHub_lock 'Microsoft.Authorization/locks@2016-09-01' = if (lock != 'NotSpecified') {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:31]:
```suggestion
resource privateLinkHub_lock 'Microsoft.Authorization/locks@2016-09-01' = if (lock != 'NotSpecified') {
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  name: '${synapsePrivateLinkHub.name}-${lock}-lock'
+  properties: {
+    level: lock
+    notes: (lock == 'CanNotDelete') ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: synapsePrivateLinkHub
+}
+
+// RBAC
+module synapsePrivateLinkHub_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, index) in roleAssignments: {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:41]:
```suggestion
module privateLinkHub_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, index) in roleAssignments: {
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  name: '${deployment().name}-rbac-${index}'
+  params: {
+    principalIds: roleAssignment.principalIds
+    roleDefinitionIdOrName: roleAssignment.roleDefinitionIdOrName
+    resourceId: synapsePrivateLinkHub.id
+  }
+}]
+
+// Private Endpoints
+module synapsePrivateLinkHub_privateEndpoints '.bicep/nested_privateEndpoint.bicep' = [for (privateEndpoint, index) in privateEndpoints: {
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:51]:
```suggestion
module privateLinkHub_privateEndpoints '.bicep/nested_privateEndpoint.bicep' = [for (privateEndpoint, index) in privateEndpoints: {
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  params: {
+    privateEndpointResourceId: synapsePrivateLinkHub.id
+    privateEndpointVnetLocation: reference(split(privateEndpoint.subnetResourceId, '/subnets/')[0], '2020-06-01', 'Full').location
+    privateEndpointObj: privateEndpoint
+    tags: tags
+  }
+}]
+
+@description('The resource ID of the deployed Synapse Private Link Hub.')
+output synapsePrivateLinkHubResourceId string = synapsePrivateLinkHub.id
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:62]:
```suggestion
output privateLinkHubResourceId string = synapsePrivateLinkHub.id
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+@description('The name of the deployed Synapse Private Link Hub.')
+output synapsePrivateLinkHubName string = synapsePrivateLinkHub.name
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:65]:
```suggestion
output privateLinkHubName string = synapsePrivateLinkHub.name
```

File: arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+@description('The resource group of the deployed Synapse Private Link Hub.')
+output synapsePrivateLinkHubResourceGroup string = resourceGroup().name
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/deploy.bicep:68]:
```suggestion
output privateLinkHubResourceGroup string = resourceGroup().name
```

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
+# Synapse PrivateLinkHubs `[Microsoft.Synapse/privateLinkHubs]`
+
+This module deploys Synapse PrivateLinkHubs.
+// TODO: Replace Resource and fill in description
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:4]: Remove

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
+# Synapse PrivateLinkHubs `[Microsoft.Synapse/privateLinkHubs]`
+
+This module deploys Synapse PrivateLinkHubs.
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:3]:
```suggestion
This module deploys a Synapse Private Link Hubs.
```

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+| Parameter Name | Type | Default Value | Possible Values | Description |
+| :-- | :-- | :-- | :-- | :-- |
+| `location` | string | `[resourceGroup().location]` |  | Required. The geo-location where the resource lives. |
+| `lock` | string | `NotSpecified` | `[CanNotDelete, NotSpecified, ReadOnly]` | Optional. Specify the type of lock. |
+| `name` | string |  |  | Required. The name of the Private Link Hub. |
+| `privateEndpoints` | array | `[]` |  | Optional. Configuration Details for private endpoints. |
+| `roleAssignments` | array | `[]` |  | Optional. Array of role assignment objects that contain the 'roleDefinitionIdOrName' and 'principalId' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11' |
+| `tags` | object | `{object}` |  | Optional. Tags of the resource. |
+
+### Parameter Usage: `<ParameterPlaceholder>`
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:27]: Remove this placeholder example

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        "Role": "DeploymentValidation"
+    }
+}
+```
+
+## Outputs
+
+| Output Name | Type | Description |
+| :-- | :-- | :-- |
+| `synapsePrivateLinkHubName` | string | The name of the deployed Synapse Private Link Hub. |
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:110]:
```suggestion
| `privateLinkHubName` | string | The name of the deployed Synapse Private Link Hub. |
```

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+| `synapsePrivateLinkHubResourceGroup` | string | The resource group of the deployed Synapse Private Link Hub. |
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:111]:
```suggestion
| `privateLinkHubResourceGroup` | string | The resource group of the deployed Synapse Private Link Hub. |
```

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+| `synapsePrivateLinkHubResourceId` | string | The resource ID of the deployed Synapse Private Link Hub. |
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:112]:
```suggestion
| `privateLinkHubResourceId` | string | The resource ID of the deployed Synapse Private Link Hub. |
```

**Review by renepajta — COMMENTED** [review]
Requested changes resolved.

File: .azuredevops/modulePipelines/ms.synapse.privatelinkhubs.yml [hunk 21]
```diff
@@ -0,0 +1,63 @@
+name: 'Storage - StorageAccounts'
```
Comment (MrMCake) [inline .azuredevops/modulePipelines/ms.synapse.privatelinkhubs.yml:1]: Update to 
```suggestion
name: 'Synapse- PrivateLinkHubs'
```

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+| `Microsoft.Synapse/privateLinkHubs`                       | 2021-06-01         |
+
+## Parameters
+
+| Parameter Name     | Type   | Default Value                | Possible Values                          | Description                                                                                                                                                                                                                                                                                                                                                                                                    |
+| :----------------- | :----- | :--------------------------- | :--------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
```
Comment (MrMCake) [inline arm/Microsoft.Synapse/privateLinkHubs/readme.md:18]: I guess you have the markdown extension set to autoformat tables. Please undo the excess spaces and '-'. These files become unreadable if you don't open them locally

Comment (MrMCake) [reply arm/Microsoft.Synapse/privateLinkHubs/readme.md:18]: There is a settings for that in the extension

File: arm/README.md [hunk 23]
```diff
@@ -2,88 +2,88 @@ In this section you can find useful information regarding the Modules that are c
...
-| [API Connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/connections) | `MS.Web` | [connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/connections) |
-| [App Service Environments](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/hostingEnvironments) |  | [hostingEnvironments](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/hostingEnvironments) |
-| [App Service Plans](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/serverfarms) |  | [serverfarms](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/serverfarms) |
-| [Web/Function Apps](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |  | [sites](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |
+| Name                                                                                                                                 | Provider namespace          | Resource Type                                                                                                                       |
```
Comment (MrMCake) [inline arm/README.md:5]: Undo all changes

File: arm/Microsoft.Synapse/privateLinkHubs/readme.md (hunk 22 above)
Comment (renepajta) [reply arm/Microsoft.Synapse/privateLinkHubs/readme.md:18]: Can we add that setting to .vscode settings.json file?

**Review by MrMCake — APPROVED** [review]

**Pull request merged**